### PR TITLE
Accept `features` in `AutoMLSearch` and add `DFSTransformer` to AutoML pipelines

### LINKED ISSUE #2919: Add `DFSTransformer` as part of AutoML when `features` are accepted as input
Add the DFSTransformer component to the beginning of EvalML pipelines by adding features as input for  AutoMLSearch() and will then be added to pipeline creation via automl algorithms

```
class AutoMLSearch:
    def __init__(
        self,
        X_train=None,
        y_train=None,
        ...,
        features=None
    )
    
    self.features = features
    if _automl_algorithm == "iterative":
            self._automl_algorithm = IterativeAlgorithm(
                max_iterations=self.max_iterations,
                allowed_pipelines=self.allowed_pipelines,
                tuner_class=self.tuner_class,
                random_seed=self.random_seed,
                n_jobs=self.n_jobs,
                number_features=self.X_train.shape[1],
                pipelines_per_batch=self._pipelines_per_batch,
                ensembling=run_ensembling,
                text_in_ensembling=text_in_ensembling,
                pipeline_params=parameters,
                custom_hyperparameters=custom_hyperparameters,
                features=self.features,
            )
        elif _automl_algorithm == "default":
            self._automl_algorithm = DefaultAlgorithm(
                X=self.X_train,
                y=self.y_train,
                problem_type=self.problem_type,
                sampler_name=self._sampler_name,
                tuner_class=self.tuner_class,
                random_seed=self.random_seed,
                pipeline_params=parameters,
                custom_hyperparameters=self.custom_hyperparameters,
                text_in_ensembling=text_in_ensembling,
                features=self.features
            )
```

### PR DESCRIPTION
**Pull request by jeremyliweishih**
Fixes #2919.
graphviz

### CODE REVIEW

File: evalml/tests/automl_tests/test_default_algorithm.py [hunk 1]
```diff
@@ -727,3 +728,69 @@ def test_default_algorithm_time_series_known_in_advance(
     long_estimators = set([pipeline.estimator.name for pipeline in long_2])
     assert len(long_2) == 30
     assert len(long_estimators) == 3
+
+
+@pytest.mark.parametrize("split", ["numeric-only", "categorical-only"])
```
Comment (jeremyliweishih) [inline evalml/tests/automl_tests/test_default_algorithm.py:733]: Still need to figure out implementation and test for the "split" pipeline case. It'll be a little more confusing as I would need to figure out how to do the follow pipeline structure:

                               /  ---------------[non-categorical pipeline]----------------------\
[DFSTransformer]                                                                                                               [estimator]
                               \ ---------------[categorical pipeline]---------------------------/

File: evalml/pipelines/utils.py [hunk 2]
```diff
@@ -657,6 +667,11 @@ def _make_new_component_name(name, component_name, idx=None, pipeline_name=None)
             new_component_name = _make_new_component_name(
                 component_pipeline_name, name, name_idx, sub_pipeline_name
             )
+            first_x_component = (
+                pipeline.component_graph.compute_order[0]
+                if pipeline.component_graph.compute_order[0] != "Label Encoder"
```
Comment (jeremyliweishih) [inline evalml/pipelines/utils.py:672]: AFAIK this is the only case where a component will come out of `make_pipelines` with a component that alters y as the very first component. I could add more logic here but that would involve instantiating a component and checking if it affects x or y.

File: evalml/pipelines/utils.py [hunk 3]
```diff
@@ -657,6 +669,11 @@ def _make_new_component_name(name, component_name, idx=None, pipeline_name=None)
             new_component_name = _make_new_component_name(
                 component_pipeline_name, name, name_idx, sub_pipeline_name
             )
+            first_x_component = (
```
Comment (freddyaboulton) [inline evalml/pipelines/utils.py:672]: I think there might be a bug in the split pipeline creation.

```python
from evalml.demos import load_fraud
from evalml.automl import AutoMLSearch
from evalml.pipelines.components import DFSTransformer
from evalml.exceptions import AutoMLSearchException
import pytest

X, y = load_fraud(1000)

dfs = DFSTransformer()

# Case 1: Including identity feature in X_train
X_t = dfs.fit_transform(X)
X_t.ww['datetime'] = X['datetime']
automl = AutoMLSearch(X_t, y, "binary", _automl_algorithm="default",
                      max_batches=4, features=dfs.features, verbose=True)

with pytest.raises(Exception, match="All pipelines in the current AutoML batch"):
    automl.search()
    
# Case 2: Not including identity feature in X_train
automl = AutoMLSearch(X, y, "binary", _automl_algorithm="default",
                      max_batches=4, features=dfs.features, verbose=True)
with pytest.raises(Exception,
                   match="not found in DataFrame",
                   ):
    automl.search()
```

These two cases pass for the iterative algorithm.

Comment (jeremyliweishih) [reply evalml/pipelines/utils.py:672]: thanks! will take a look, still stuck on running integration tests so I'm sure theres more breaking!

File: evalml/tests/automl_tests/test_automl.py [hunk 4]
```diff
@@ -4485,3 +4486,46 @@ def test_cv_validation_scores_time_series(
...
+        features=features,
+        automl_algorithm=automl_algorithm,
+    )
+    assert automl.automl_algorithm.features == features
+    env = AutoMLTestEnv("binary")
+    with env.test_context(score_return_value={automl.objective.name: 1.0}):
+        automl.search()
+    assert all(
+        [
+            p["DFS Transformer"]["features"] == features
```
Comment (chukarsten) [inline evalml/tests/automl_tests/test_automl.py:4528]: For completeness, it might be good to add in the alternative test case, where if features is None or not provided, DFS transform shouldn't be in there.  But this is a good test and obvious as to what you're testing.  I like parameterizing by switching on a string in the pytest decorator so that the test output in pytest or pycharm says things like "with_features_provided" and "without_features_provided".

File: evalml/tests/pipeline_tests/test_pipeline_utils.py [hunk 5]
```diff
@@ -986,3 +986,65 @@ def test_make_pipeline_from_multiple_graphs_with_sampler(X_y_binary):
         combined_pipeline.component_graph.get_inputs("Random Forest Classifier")[2]
         == second_pipeline_sampler
     )
+
+
+def test_make_pipeline_from_multiple_graphs_prior_components(X_y_binary):
```
Comment (chukarsten) [inline evalml/tests/pipeline_tests/test_pipeline_utils.py:991]: Cool, nice test, very obvious what you're testing.

**Review by chukarsten — APPROVED** [review]
Left a few questions, mostly for my education and understanding, but this looks good and looks like you accomplished what you set off to do!

File: evalml/automl/automl_algorithm/default_algorithm.py [hunk 6]
```diff
@@ -70,6 +70,7 @@ class DefaultAlgorithm(AutoMLAlgorithm):
         num_long_pipelines_per_batch (int): number of pipelines per batch for each top n pipeline through long mode.
         allow_long_running_models (bool): Whether or not to allow longer-running models for large multiclass problems. If False and no pipelines, component graphs, or model families are provided,
             AutoMLSearch will not use Elastic Net or XGBoost when there are more than 75 multiclass targets and will not use CatBoost when there are more than 150 multiclass targets. Defaults to False.
+        features (list)[FeatureBase]: List of features to run DFS on in AutoML pipelines. Defaults to None. Features will only be computed if the columns used by the feature exist in the input and if the feature itself is not in input.
```
[the remark this answers is not in the document]
Comment (jeremyliweishih) [reply evalml/automl/automl_algorithm/default_algorithm.py:73]: I'm trying to convey that if `ABSOLUTE(1)` was passed in as a feature, it will only be computed if:
1. `1` is a column in the input
2. `ABSOLUTE(1)` doesn't already exist in the input

Will def try to reword though!

File: evalml/pipelines/utils.py [hunk 7]
```diff
@@ -644,11 +652,15 @@ def _make_new_component_name(name, component_name, idx=None, pipeline_name=None)
     parameters = copy.deepcopy(parameters) if parameters else {}
     final_components = []
     used_names = []
-    component_graph = (
-        {"Label Encoder": ["Label Encoder", "X", "y"]}
-        if is_classification(problem_type)
-        else {}
+
+    prior_components = {} if not prior_components else prior_components
```
[the remark this answers is not in the document]
Comment (jeremyliweishih) [reply evalml/pipelines/utils.py:689]: most definitely but I'm trying to avoid having an empty dictionary as a parameter. But since we don't return `prior_components` it should be safe to do so. Will update!

File: evalml/pipelines/component_graph.py [hunk 8]
```diff
@@ -366,6 +366,14 @@ def predict(self, X):
         )
         return infer_feature_types(outputs.get(f"{final_component}.x"))
 
+    def _return_non_engineered_features(self, X):
+        base_features = [
+            c
+            for c in X.ww.columns
+            if X.ww[c].ww.origin == "base" or X.ww[c].ww.origin is None
```
[the remark this answers is not in the document]
Comment (jeremyliweishih) [reply evalml/pipelines/component_graph.py:376]: yup! Featuretools will tag engineered features as "engineered" and "base" otherwise. https://featuretools.alteryx.com/en/latest/resources/transition_to_ft_v1.0.html#Changes-to-Deep-Feature-Synthesis-and-Calculate-Feature-Matrix
